packages/core/admin/server/src/services/homepage.ts

```diff
@@ -1,129 +1,212 @@
-          if (mainField) {
-            fields.push(mainField);
-          }
+  type DocumentMeta = {
+    fields: string[];
+    mainField: string;
+    contentType: Schema.ContentType;
+    hasDraftAndPublish: boolean;
+    uid: RecentDocument['model'];
+  };
+
+  const getDocumentsMetaData = (
+    allowedContentTypeUids: RecentDocument['model'][],
+    configurations: ContentTypeConfiguration[]
+  ): DocumentMeta[] => {
+    return allowedContentTypeUids.map((uid) => {
+      const configuration = configurations.find((config) => config.uid === uid);
+      const contentType = strapi.contentType(uid);
+      const fields = ['documentId', 'updatedAt'];
+
+      // Add fields required to get the status if D&P is enabled
+      const hasDraftAndPublish = contentTypes.hasDraftAndPublish(contentType);
+      if (hasDraftAndPublish) {
+        fields.push('publishedAt');
+      }
+
+      // Only add the main field if it's defined
+      if (configuration?.settings.mainField) {
+        fields.push(configuration.settings.mainField);
+      }
+
+      // Only add locale if it's localized
+      const isLocalized = (contentType.pluginOptions?.i18n as any)?.localized;
+      if (isLocalized) {
+        fields.push('locale');
+      }
+
+      return {
+        fields,
+        mainField: configuration!.settings.mainField,
+        contentType,
+        hasDraftAndPublish,
+        uid,
+      };
+    });
+  };
+
+  const formatDocuments = (documents: Modules.Documents.AnyDocument[], meta: DocumentMeta) => {
+    return documents.map((document) => {
+      const { mainField, ...restDocument } = document;
+
+      return {
+        data: {
+          ...restDocument,
+          ...(restDocument.publishedAt && { publishedAt: new Date(document.publishedAt) }),
+          updatedAt: new Date(document.updatedAt),
+          title: document[mainField ?? 'documentId'],
+        },
+        meta: {
+          model: meta.uid,
+          kind: meta.contentType.kind,
+          hasDraftAndPublish: meta.hasDraftAndPublish,
+        },
+      };
+    });
+  };
 
-          // Only add locale if it's localized
-          const isLocalized = (contentType.pluginOptions?.i18n as any)?.localized;
-          if (isLocalized) {
-            fields.push('locale');
+  const addStatusToDocuments = async (
+    documents: {
+      data: RecentDocument[];
+      meta: {
+        model: RecentDocument['model'];
+        kind: RecentDocument['kind'];
+        hasDraftAndPublish: boolean;
+      };
+    }[]
+  ): Promise<RecentDocument[]> => {
+    return Promise.all(
+      documents.map(async (document) => {
+        /**
+         * Tries to query the other version of the document if draft and publish is enabled,
+         * so that we know when to give the "modified" status.
+         */
+        const { availableStatus } = await metadataService.getMetadata(
+          document.meta.model,
+          document.data,
+          {
```

Comment:
Maybe you can extend your DocumentMeta type here?

Reply:
this ended up in the wrong place I meant there: https://github.com/strapi/strapi/pull/22377/files#diff-eb8ec54830d7c9a5050f556298d317736043775b05e26620d260fde437253b79R118-R122